### ISSUE TITLE
cXXX6 - Improved `__attribute__((cleanup))` Through `defer`

### ISSUE THREAD

**Issue by ThePhD** [ThePhD/future_cxx#67]

**Comment by ThePhD** [comment]
It took a while, but the paper was fully written.

Current standard revision: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3199.htm

**Comment by jiyuzh** [comment]
Thank you for the proposal. I like it very much! I have one question regarding Section 3.8.3: It seems there is no specification over whether `break` or `continue` are allowed to be used to jump over a `defer`. Do you think this type of jump over should be allowed or disallowed?

**Pros**

Consider the following use case:

```c
for (int i = 0; i < 10; i++) {
    void *ptr = malloc(8);

    if (!ptr)
        break;

    defer { free(ptr); }
}
```

It seems we should allow such jump over in this example. Also, `break` is semi-related to `switch`, allowing `break` to jump over seems making loops and `switch` consistent on the constraint.

**Cons**

Some common `goto` usages can also be expressed with/converted to loops plus `break` or `continue`, hence if `goto` is banned, so shall `break` and `continue`. The wording of the draft also seems to imply `return` is an exception. Also, if labeled loops are adopted into C standard, the jump over could be even more error prone.

**Personal opinion**

I implemented a C-based polyfill using `__attribute__((cleanup(...)))` and used it in several of my projects.

During my usage, I find `goto` and `switch`-jump-over are indeed confusing. However, based on my personal feeling, the `break`/`continue` use case could be a good thing to have.

```c
#define _DEFER_MERGE(a,b) a##b
#define _DEFER_VARNAME(a) _DEFER_MERGE(____defer_scopevar_, a)
#define _DEFER_FUNCNAME(a) _DEFER_MERGE(____defer_scopefunc_, a)
#define _DEFER(n) \
    auto void _DEFER_FUNCNAME(n)(int *a); \
    __attribute__((cleanup(_DEFER_FUNCNAME(n)))) int _DEFER_VARNAME(n); \
    void _DEFER_FUNCNAME(n)(int *a)

#define defer _DEFER(__COUNTER__)
```

It closely resembles some behaviors defined by the draft (i.e., example 2, 4, 5 from Section 6.4), which can be tested here: https://godbolt.org/z/x7eos13T5

**Comment by ThePhD** [comment]
Thanks for your support!

Regarding jumps like `goto`, `break`, and `continue`, the latest version of the proposal has the following constraints in the proposed wording:

> **Constraints**
> 
> Jumps by means of `goto` or `switch` into *E* shall not jump over a `defer` statement in *E*.
> 
> Jumps by means of `goto` or `switch` shall not jump into any `defer` statement.
> 
> Jumps by means of `return`, `goto`, `break`, or `continue` shall not exit *S*.

*S* here is the scope of the `defer` statement itself. So the only jumps that are banned are ones that jump **out of** a `defer`, not ones that happen to skip its execution. Your example:

```c
for (int i = 0; i < 10; i++) {
    void *ptr = malloc(8);

    if (!ptr)
        break;

    defer { free(ptr); }
}
```

is perfectly fine and valid. if `ptr` is `NULL`/a null pointer, then this code will simply exit the loop and the `defer` will not be called because it has not been reached in that scope yet. We do not ban any use of `goto` or friends that is appropriate, only banning if it jumps **over** a defer statement. You can `goto` into some block/scope that contains a `defer`, so long as you don't jump over it:

```c
int handle = get_handle();
goto meow;  // label after defer: using `goto` to jump forwards or backwards over is illegal
defer { release_handle(handle); }
meow0; ;
meow1: ;
printf("hiiii"); 
goto meow1; // okay
```

I think that covers the cases you were speaking of. Also, even if someone implements `break` and `continue` like a `goto`, that's not of the C Standard's concern. The examples and informative (not normative) explanatory text demonstrating it in terms of `goto`s is not a mandate for the implementation to follow: a C compiler (or interpreter, or transpiler, or any C implementation) does not have to care about such hints. They have to uphold the **Constraints** and **Semantics** sections, as given. That most flow control can be flattened/rewritten in terms of `goto` is not something the C standard has to care about, so long as the exact use of `continue` or `break` follow what is in the standard before your compiler changes it / reorders it / messes with it.

**Comment by BatmanAoD** [comment]
One small organizational issue: the last paragraph of 3.4 (all of which is duplicated in N3198) states that "for the macros we provide almost every single one will be defined and have the value of `1`", which makes sense in the context of N3198 but not N3199. N3198 is linked later, in 3.4.2; if N3198 were mentioned and linked prior to the note about MSVC and SEH, then the note could allude to "the macros provided in N3198."

**Comment by tylov** [comment]
Hi. First, I think this is an excellent paper, and I agree with basically all the reasoning in this proposal. 

Regarding the **Library-based Solutions** section, [I made an implementation](https://github.com/stclib/STC/blob/v50dev/include/stc/algo/defer.h) that I think has some significant advantages over moon-chilled's implementation, which I compare with as it is highlighted in the proposal. 

For the committee to best evaluate the limitations of **defer** library solutions, it is useful for them to have state-of-art implementations available, and it is where my implementation may have a place. These will also serve programmers best until **defer** is standardized, supported by compilers, and until people are ready to commit to the C-standard which supports it. It is also an advantage if the library solution usage is close to how the accepted standard will likely look like, for easy migration.

A short comparison of mine versus moon-chilled's implementation:
| Mine (STC) | Moon-chilled
|:----------------|:----------------
| Tiny code, easy to deploy | Tiny code, easy to deploy
| Highly efficient, **all portable** code | Highly efficient non-portable- / Slow, portable code
| Natural "guard" scope syntax: `c_scope {... }` | Function start macro, required **Return** for void funcs.
| Two nesting levels: `c_scope {... c2_scope {... }... }` | Function-level scope only
| Multiple defer scopes per function with `cs_scope {... }` | Function-level scope only
| Break out of a defer scope with `continue` | N/A

`c2_scope` and `cs_scope` uses _setjmp_, however they are much less used. `c_scope` is the primary fast (_switch_-based) defer scope.

**Comment by tylov** [comment]
> There are, as far as most reviewers can tell, no errors in the creation or deletion of the various kinds of resources (particularly, repeated memory allocations).

**Most** reviewers? Each of the returns inside the loop will leak namelist in addition to all namelist[k]'s except for k in [0, i) where namelist[k]->d_name[0] == '.'

**Comment by tylov** [comment]
If the same programmer had access to defer, it would still go wrong unless he/she realized that it had to be rewritten like you did.
```c
  struct dirent** namelist;
  int n = scandir(pluginsdir, &namelist, NULL, alphasort);
  defer free(namelist);
  for (int i = 0; i < n; ++i)
  {
    struct dirent* ent = namelist[i];
    defer free(ent); // <= mem-leak if there are early returns when i < n-1
    if (ent->d_name[0] == '.')
    {
      continue;
    }
    ...
```

**Comment by fuhsnn** [comment]
Greetings, if implementation experience from an amateur compiler could help get this through, I wrote a little bit [here](https://github.com/fuhsnn/slimcc-defer#implementation-experience) along with the implementation.

**Comment by fuhsnn** [comment]
Some feedback on C++ compatibility (hope this is relevant, since the draft does contain sizable words for that):

The lambda based scoped guard implementation under `4.4. The Polyfill/C++ Fix` may not behave as expected (as my understanding of the proposed C defer) when copy-elision is in effect, for example `EXAMPLE 4` converted to struct may return 5 instead of 4 on all major C++ compilers: https://godbolt.org/z/q41rMo5Ex

My short discussion with LLVM people: https://github.com/llvm/llvm-project/issues/100869

**Comment by ThePhD** [comment]
> The lambda based scoped guard implementation under `4.4. The Polyfill/C++ Fix` may not behave as expected (as my understanding of the proposed C defer) when copy-elision is in effect, for example `EXAMPLE 4` converted to struct may return 5 instead of 4 on all major C++ compilers: https://godbolt.org/z/q41rMo5Ex

Nothing I can do about that, unfortunately. Maybe if C gets a form of applicable NRVO, but otherwise the intent and the effect is still matching and the same.

**Comment by ThePhD** [comment]
> Greetings, if implementation experience from an amateur compiler could help get this through, I wrote a little bit [here](https://github.com/fuhsnn/slimcc-defer#implementation-experience) along with the implementation.

Thank you for the implementation. At the moment `defer` is likely headed to a Techincal Specification unless we can convince folks that it should just go straight into the next working draft instead. I might do that but I'm tired of fighting people on this stuff so I'm just writing a Technical Specification; I will link to and include your work in the next revision of the proposal, however.

**Comment by fuhsnn** [comment]
Thanks, just keep doing what you're doing. I'll keep stalking wg14 documents for fun stuff to implement.

**Comment by strncmp** [comment]
Hello, I just wanted to ask if any provisions to *completely* solve the resource leak and dangling resource problems in C were mentioned by the committee during the discussions of this feature. I know this feature is supposed to help with that, and I appreciate it very much, but it doesn't protect the user in a language level manner (you still need "better analysers and linters" and many reviewers to gain confidence about your resource use, and you still cannot be sure that you haven't missed anything). That being said, thank you for working on this, it's really useful.

**Comment by ThePhD** [comment]
C cannot fundamentally solve this problem without changing core parts of C.

Have you considered writing Rust?

**Comment by ThePhD** [comment]
If you'd like to know the Committee Strategy, you can see the Charter that was updated for C2y, which added a much higher focus on application security: https://www.open-std.org/JTC1/SC22/WG14/www/docs/n3280.htm

Nevertheless, the C committee cannot cook up features. People, like me, have to write proposals. They can be coordinated, but generally they are done independently. Other people are trying their own things which they have shown interest in standardizing but have written no papers yet, such as:
https://discourse.llvm.org/t/rfc-enforcing-bounds-safety-in-c-fbounds-safety/70854 
https://llvm.org/devmtg/2023-05/slides/TechnicalTalks-May11/01-Na-fbounds-safety.pdf
https://www.youtube.com/watch?v=RK9bfrsMdAM
https://clang.llvm.org/docs/BoundsSafety.html

**Comment by ryansuchocki** [comment]
Thanks for all the work that this must have taken. It's exciting to see that things like this are on the cards.

I'm curious about the rationale for proposing the `defer {<cleanup code>} <protected code>` structure rather than something like `do {<protected code>} finally {<cleanup code>}`.

The `do {...} finally {...}` structure would seem dramatically more intuitive to me as a piece of C syntax, but moreover I think it has some distinct advantages which would make it easier to teach and explain the semantics (especially when it comes to the interactions with `return`/`break`/`continue`/`goto`):
* The code appears in the order in which it runs (i.e. the cleanup code appears after the "protected" code)
* The "protected code" is identified with explicit lexical structure (either braces or the space between the keyword pair) rather than being "everything from here to the end of the enclosing block"

I'm sure this has had consideration, so I'd like to hear your perspective. I think I saw in the minutes of one of the meetings that there had been some debate on this aspect of the proposal (but there was no detail).

**Comment by tylov** [comment]
There is another option, which is `__try { ... __finally: ... }`. This has the same benefits of appearing in the order in which it runs, but still within a limited scope. It is also intuitive because the code flows naturally into the `__finally:` stage unless a `return` already sent it there before returning.  Was this considered?

**Comment by ThePhD** [comment]
This was not considered, and I personally won't be considering it because I don't think having a special label is good enough. This does not handle `goto` or `switch` or `break/continue` leaving a given scope.

**Comment by ThePhD** [comment]
The TS is done and will go through the usual hoop-jumping process in a bit: https://www.open-std.org/JTC1/SC22/WG14/www/docs/n3489.pdf

**Comment by fdwr** [comment]
💚 Thank you pushing through this long-overdue missing control flow construct. I agree with all the principles (e.g. using the value at the time of actual execution consistently with other control flow blocks rather than some unclear capture logic, nesting, reverse unwinding order, no goto in to/out from...), and it feels quite intuitive (unlike Go's `defer`) and harmonious with the existing C grammar. I *briefly* questioned whether a *dismissible* defer might be needed (in case a success path cleans up sooner, like clearing an `std::optional` early), but then realized implementing that is as simple as adding a boolean local and `if` statement in the `defer` block.

`defer` would be super useful for C++ too (C++32? 😅🤞) for a lot of those one-off cleanup cases where it feels silly to write a whole mini-helper class, like:

```cpp
RETURN_IF_FAILED(CoInitializeEx(nullptr, COINIT_MULTITHREADED));
defer CoUninitialize();

// ... more stuff ...
```

Some might question, "Well if we have C++ destructors, then why would we need defer", without realizing that (a) there are many kinds of cleanup and post-action that don't fit into standard destructor semantics (b) if you already had the `defer` block concept, that's conceptually that's how destructors could have been implemented. So `defer` complements `destruct`.

> My only worry is that Microsoft is going to ... ignore ... any forward progress with just a defer TS. [(quote)](https://thephd.dev/c2y-the-defer-technical-specification-its-time-go-go-go)

Note there is a strong security push inside Microsoft lately for OS components ([Secure Future Initiative](https://learn.microsoft.com/en-us/windows/security/book/security-foundation-offensive-research)), and so anything positioned from that angle has more weight (and [resource leaks can indeed cause CVEs](https://www.cve.news/cve-2024-53084/), as [you mentioned](https://thephd.dev/c2y-the-defer-technical-specification-its-time-go-go-go#cve-2021-3744-and-the-truth-about-programmers)).

**Comment by leni536** [comment]
> I opened a VS [feature request here](https://developercommunity.visualstudio.com/t/Support-defer-cleanup-blocks-TS/10874355).

Are there similar GCC and LLVM tickets?

**Comment by asd142513** [comment]
I've written up an alternative approach to the proposed defer-statement.
Any comments would be appreciated!

```c
/*
Original post:
https://thephd.dev/c2y-the-defer-technical-specification-its-time-go-go-go

Inspired by for loop and Python’s `with` statement:
https://en.cppreference.com/w/c/language/for.html
	for ( init-clause ; cond-expression ; iteration-expression ) loop-statement
https://docs.python.org/3/reference/compound_stmts.html#with

Syntax:
	with ( init-clause ; cond-expression ; cleanup-expression ) context-statement

Comparison: `thephd.dev defer` vs `with`
	1. Automatically creates a new scope.
	2. Supports only expressions — no deferred statements.
	3. `break` is allowed within the `with` block.
	4. Nested `with` statements are not allowed within init-clause, cond-expression, or cleanup-expression.
	   However, the context-statement can include `with`, just like in loop bodies.
	5. No mid-scope defers.
	   - In the original `defer`, deferred statements could appear anywhere in the scope.
	   - This is suboptimal, as developers must read the entire block to find all deferred actions.
	   - The `with` statement improves clarity by forcing all deferred behavior to be declared at the start of the block.
	6. Unlike `defer`, which helps maintain a flat block structure `with` introduces an additional level of nesting.
	   This is even worse in common `malloc`–`init_object` patterns as it introduces two levels of additional nesting.

Comparison: `for` loop vs `with` block
	1. Trivially, context-statement is executed only once; `continue` is not allowed.
	2. Unlike a `for` loop's iteration-expression, the cleanup-expression is evaluated when exiting via `break` or `return`.
	   Jump to outside of block with `goto` skips the cleanup-expression.
*/

#include <threads.h>

extern int do_sync_work(int id, mtx_t *m);

int main(void)
{
	with (mtx_t m = {}; mtx_init(&m, mtx_plain) == thrd_success; mtx_destroy(&m)) {
		for (int i = 0; i < 12; ++i) {
			with (; mtx_lock(&m) == thrd_success; mtx_unlock(&m)) {
				if (do_sync_work(i, &m) == 0) {
					return 1;
				}
			}
		}
	}

	return 0;
}
```
[gist link](https://gist.github.com/asd142513/ee494971fc93b46a4ed81935e2e4261a)

Apologies if this isn't the appropriate place to share such alternatives.

English is not my first language. All content was written by hand and then refined using AI. I'd appreciate it if you could point out any awkward phrasing.

**Comment by ThePhD** [comment]
@leni536 -- GCC has an open patch going, with newer e-mails doing more. https://gcc.gnu.org/pipermail/gcc-patches/2025-August/692757.html

Haven't spotted anything in Clang/LLVM (yet).

**Comment by fdwr** [comment]
> GCC has an open patch going

I'd love to try it on [Godbolt](https://godbolt.org/) as soon as any trunk/experimental builds are available along with any experimental command line parameters needed to enable it. ⏳🤞

**Comment by fdwr** [comment]
> GCC has an open patch going ... Haven't spotted anything in Clang/LLVM (yet).

## Updates:
- ✅ **Clang** trunk supports `defer` now. See https://godbolt.org/z/dxhnGb7nf:
  - **2025-12-10**: Initial change https://github.com/llvm/llvm-project/pull/162848.
  - **2025-12-16**: stddefer.h fix https://github.com/llvm/llvm-project/pull/172512
- ✅ **Cake** (C23 compiler frontend) supports `_Defer`: http://cakecc.org/manual.html#toc_69
  - **2025-12-03**: Align with n3734: https://github.com/thradams/cake/commit/cf1253b80afaa28bd361b3c0dc50751effe8b860
- ⏳**GCC**
  - **2025-01-02**: It's evidently not on any https://godbolt.org/z/da8x1r8E3 trunk/experimental builds yet ⏳ (tried 15.2 and trunk with `-std=c2y`)
  - **2025-10-08**: `-fdefer-ts` should enable it *once* in trunk per https://gcc.gnu.org/pipermail/gcc-patches/2025-October/696666.html. Currently we get `cc: error: unrecognized command-line option '-fdefer-ts'...`.
  - **2025-12-11**: `_Defer` is provided as a keyword if gnu extensions are enabled, and stddefer.h provided a macro for the 'defer' spelling https://gcc.gnu.org/pipermail/gcc-patches/2025-December/702875.html
  - **2026-01-14**: `_Defer` is enabled by default. 'defer' is gated by '-fdefer-ts'. https://gcc.gnu.org/pipermail/gcc-patches/2026-January/705778.html
  - *Nothing new yet - awaiting to see if April brings updates. https://gcc.gnu.org/pipermail/gcc-patches/2026-April/.*
